### ISSUE TITLE
Support for Flutter extension

### ISSUE THREAD

**Issue by CodeWizard0367** [zed-industries/extensions#247]
### Check for existing issues

- [X] Completed

### Describe the feature

Please bring support for Flutter extension

**Comment by JosephTLyons** [comment]
I've moved this issue over to the extensions repository.

**Comment by CodeWizard0367** [comment]
Anyone can explain, what would it take to support languages or if someone wants to start on it, where they can start?

**Comment by adamglin0** [comment]
> Anyone can explain, what would it take to support languages or if someone wants to start on it, where they can start?

zed doesn't have a plugin system yet, so I think only the zed team can make the Flutter Plugin.

**Comment by erlangparasu** [comment]
waiting this feature

**Comment by jaswantk917** [comment]
Yes please

**Comment by followthemoney1** [comment]
lemme know if you need help on this `🦾

**Comment by ziadsarour** [comment]
Would be really awesome !

**Comment by Chandram-Dutta** [comment]
imo a flutter support would need

- detecting/opening & choosing simulators 
- compatible with flutter run scripts
- flutter dev tools
- hot reload/restarts on saving a file

I dont think zed's present plugin system really can be made extensible in this way, would like have to bake these features into the editor

**Comment by WhiteKr** [comment]
About Flutter features, hot-reload should be the first-priority one to ship into Zed in sooner. PLEASE!

**Comment by MaximilianFlechtner** [comment]
+1

**Comment by GioPan04** [comment]
Please, stop posting useless comments like "+1" or "please add this, I need this", that's not going to help, and only add spam to everyone's inbox.
Just add a "👍" to the original comment.

Have a great day y'all

**Comment by cavin-7span** [comment]
Waiting for this one

**Comment by hiteshpatel1992** [comment]
any update on flutter support ?

**Comment by justkelvin** [comment]
I am really waiting for this guys, the reason i decided to try out Zed was because VS Code was using up all my resources working on large projects

**Comment by yoziliks** [comment]
add Device manager

**Comment by iSaqibShafique** [comment]
Waiting for the language support

**Comment by brainzaps** [comment]
> Waiting for the language support

https://zed.dev/docs/languages/dart

**Comment by JulianBissekkou** [comment]
When looking at the [docs](https://github.com/zed-industries/extensions/blob/main/AUTHORING_EXTENSIONS.md) I have the feeling that we might not be able to create a flutter extension that is similar to the current VSCode or IntelliJ extensions since we are not able to customize the UI with additional buttons or add a new setting panel where users can define the path to the flutter and dart SDK.

We don't have to support all the features of the existing Flutter extensions, but being able to select a device in the a drop-down and click a run button would be a first draft 🤔 

Is that possible or are there existing zed plugins as an example?

**Comment by Chandram-Dutta** [comment]
> When looking at the [docs](https://github.com/zed-industries/extensions/blob/main/AUTHORING_EXTENSIONS.md) I have the feeling that we might not be able to create a flutter extension that is similar to the current VSCode or IntelliJ extensions since we are not able to customize the UI with additional buttons or add a new setting panel where users can define the path to the flutter and dart SDK.
> 
> We don't have to support all the features of the existing Flutter extensions, but being able to select a device in the a drop-down and click a run button would be a first draft 🤔 
> 
> Is that possible or are there existing zed plugins as an example?

umm so it's not possible to have the drop down as an extension. u can run the main function in the present dart extension and it works if a simulator is running or a device is selected.

the only way we can have advanced features like dropdown device selection/flutter inspector is by creating a fork of zed itself. i did try this and was able to achieve it to a level but due to my limited knowledge of rust, gpui and zed codebase, it was extremely buggy and kinda unusable.

**Comment by JulianBissekkou** [comment]
@Chandram-Dutta 
I see. Forking shouldn't be required to build UI for extensions. This is quite limiting right now.
The biggest problem I see with building extension is the lack of examples. I am willing to get into rust and gpui but we need more guidance for this. :/

**Comment by iapicca** [comment]
@Moshyfawn 
in your [message](https://github.com/zed-industries/zed/issues/5343#issuecomment-2081760538) you mention that flutter and dart won't be supported in "core zed"

at the time of writing this issue is by far the most requested in the extensions repo

<img width="1238" alt="image" src="https://github.com/user-attachments/assets/19ece128-2d71-46e8-991c-9bd7a01f95a6">

is this issue currently worked on? or there is any timeline for the beginning of the work on the issue?
thank you

**Comment by kyleorin** [comment]
would be great

**Comment by Haraven** [comment]
Hello. I'm interested in switching over from Cursor (VSCode-based IDE) but I need Flutter support before I can fully commit to Zed.

So, I'd be very happy to see Flutter support in Zed!

**Comment by MaherSafadii** [comment]
any update on this?

**Comment by OmarBakry-eg** [comment]
Any update on this?

**Comment by nemr0** [comment]
Any update on this?

**Comment by letto4135** [comment]
Why specifically is a dropdown needed on the UI? Why can't something be made that integrates with the command palette?

**Comment by md-weber** [comment]
> Why specifically is a dropdown needed on the UI? Why can't something be made that integrates with the command palette? Or something to work with the .zed settings of the project?

I guess it comes from what Flutter developers are used from VSCode and Android Studio, I could imagine if you have a better workflow that we could adapt but some things are just important. As a Flutter Developer we would love to have something similar like in VSCode where we can choose devices, and have a quick access to the main features of Flutter "Hot Reload" and "Hot Restart"

![image](https://github.com/user-attachments/assets/4ad7f919-5a6c-40d3-a1a1-b46138241920)
![image](https://github.com/user-attachments/assets/bf352866-9fe5-4ba9-b870-ae2207d19f55)
![image](https://github.com/user-attachments/assets/9cd1dc1a-31fe-4daf-94a7-54e2b57f4a53)

Also Flutter comes with the Flutter dev tools like an inspector, that is integrated in VSCode which probably a lot of people would also like to see in the IDE in the long run. So to answer your question, no a basic setting would be not sufficient because Flutter Developers are quite spoiled in other IDEs.

**Comment by letto4135** [comment]
> VSCode and Android Studio
If you want something similar to VSCode or Android Studio just use those... Zed isn't either of those and likely won't be, so either you change your methods to suit the editor or you're not likely to get the feature you want. Idk who the Zed contributors are, but would be interested to hear if their opinion aligns or if they think there is room for a change on Zeds end.

**Comment by WhiteKr** [comment]
> If you want something similar to VSCode or Android Studio just use those... Zed isn't either of those and likely won't be, so either you change your methods to suit the editor or you're not likely to get the feature you want. Idk who the Zed contributors are, but would be interested to hear if their opinion aligns or if they think there is room for a change on Zeds end.

How about we steer away from telling users to switch? Users typically want to have that same awesome experience here in Zed that they find elsewhere, and it’s totally normal to want the tools and features they love in their chosen work environment.
Instead of just shutting down their suggestions, let's chat about whether and how these features might fit into Zed, keeping in mind what makes the editor special but also being open to improvements. Having open and friendly discussions with users could really help build a collaborative community and guide Zed to grow in ways that meet its users’ varied needs.

**Comment by letto4135** [comment]
> Instead of just shutting down their suggestions

I'm not shutting down solutions, I'm suggesting an alternative that works with the current infrastructure and doesn't require a design change from the core Zed devs that may go against the philosophy of the project. From what it sounds like you're really just wanting what you want with no room for concessions, so as I said you may have to just use VS Code if you're not willing to concede that it may not work exactly the same way.

**Comment by WhiteKr** [comment]
> I'm not shutting down suggestions, I'm suggesting an alternative that works with the current infrastructure and doesn't require a design change from the core Zed devs that may go against the philosophy of the project and people seem to be shutting it down. From what it sounds like you're really just wanting what you want with no room for concessions, so as I said you may have to just use VS Code if you're not willing to concede that it may not work exactly the same way. Do I want better flutter support? Heck yeah. I really like flutter, but I at least understand that it may not be the same level of support that I get in another IDE, and I'm fine with that personally because other IDEs try too hard.

**Comment by WhiteKr** [comment]
> I really like flutter, but I at least understand that it may not be the same level of support that I get in another IDE, and I'm fine with that personally because other IDEs try too hard.

I agree, they try too hard.
What I mentioned earlier was in response to your previous comment:

> If you want something similar to VSCode or Android Studio just use those...

Because it might seem like "If you like it so much, why not use that instead?" But after hearing your explanation, I understand now.

As a Flutter and Zed fan, I can understand that other editors' Flutter features might not align with Zed's colour. But I can also understand other users.

Proposing new "Zed-ish" ideas is hard. We just want features we've experienced elsewhere in Zed. It's natural that suggestions are similar to those from other tools because they provide a good experience.
I believe creative people like you are willing to and will develop these suggestions and make them better! So, why don't you give it a try? I feel how much you love Zed, thank you.

**Comment by letto4135** [comment]
By the way, if anyone here does want to use zed *as is* I found a package yesterday that supports auto hot reload when running flutter from the terminal.
https://pub.dev/packages/dashmon

**Comment by HarshNarayanJha** [comment]
For hot reload, I actually use a custom solution which includes running flutter with
```fish
flutter run --pid-file /tmp/flutter.pid
```

and then another script that I run called `./hot-reload.sh`
```fish
watchfiles "kill -USR1 $(cat /tmp/flutter.pid)" ./lib/*
```
Which uses [watchfiles](https://github.com/samuelcolvin/watchfiles).

**Comment by vleeuwenmenno** [comment]
Since now Zed supports debugging, is there any chance we might soon have at least step-debugging and variable watching and such for Dart lang/Flutter?

I'm unsure what steps need to be taken to make this happen but it looks like Zed supports DAP https://microsoft.github.io/debug-adapter-protocol/

So an extension should now be feasible? https://zed.dev/docs/extensions/debugger-extensions

**Comment by ZorinFoss** [comment]
There is a [PR](https://github.com/zed-industries/zed/pull/31336) that add Dart/Flutter debug adapter for anyone want to test it

**Comment by l1qu1d** [comment]
> There is a [PR](https://github.com/zed-industries/zed/pull/31336) that add Dart/Flutter debug adapter for anyone want to test it

That PR was closed in favor of: https://github.com/zed-extensions/dart/pull/16.

**Comment by Ikxyz** [comment]
Any recent update on this

**Comment by anilaydinn** [comment]
I would love to see official Flutter support in Zed. As a Flutter developer, the speed of Zed is very appealing, but the lack of full language server support and debugging for Dart/Flutter is the only thing keeping me from switching. This would be a game-changer for the community! 🤟

**Comment by edTheGuy00** [comment]
Having switched to Zed a few months ago, I just decided to try building out my own little terminal based solution that turned into a full fledge side project.

https://github.com/edTheGuy00/fdemon

- Hot reload on file change
- Multiple configuration management
- Run multiple sessions on different devices (they all hot reload on file changes).
- Launch Simulators and emulator
- Direct DevTools integration.
- Code link highlights and selection (directly opens up in Zed)

And much more.

I use this for all my projects now, works for me, try it out if you want.

**Comment by mambucodev** [comment]
I'd like to develop a full Flutter extension for Zed, I'll work on it and eventually link a PR to this issue. I'll try to fit as many features as I can

**Comment by mambucodev** [comment]
Following up on my previous comment — I've gone through the entire thread and compiled a table of all the features that could eventually be part of a Flutter extension for Zed. This should serve as a rough roadmap.

| # | Feature | Description | Priority | Notes |
|---|---------|-------------|----------|-------|
| 1 | Hot Reload | Automatically trigger hot reload on file save | High | Most requested feature in this thread. Workarounds exist (`dashmon`, `watchfiles` + `--pid-file`, `fdemon`) but native support is needed |
| 2 | Hot Restart | Trigger a full hot restart from the editor | High | Complements hot reload for state-resetting changes |
| 3 | Device Manager | List, select, and launch simulators/emulators | High | Device selector dropdown or command palette integration |
| 4 | Run / Launch Support | Run Flutter apps from the editor with a selected device | High | Should support multiple run configurations |
| 5 | DAP Debugger Integration | Step debugging, breakpoints, variable watching | High | Zed now supports DAP — see [dart#16](https://github.com/zed-extensions/dart/pull/16) for prior work |
| 6 | DevTools Integration | Open Flutter DevTools (inspector, performance, etc.) from the editor | Medium | Could launch DevTools in browser or embed if Zed allows it |
| 7 | Flutter Commands | Access to `flutter doctor`, `flutter clean`, `flutter pub get`, etc. | Medium | Via command palette or tasks |
| 8 | SDK Path Configuration | Allow users to configure Flutter/Dart SDK paths | Medium | Needed for non-standard SDK installations |
| 9 | Multiple Run Configurations | Support for different flavors, targets, and build modes | Medium | Similar to launch.json configs in VS Code |
| 10 | Widget Inspector | Inspect widget tree, layout, and properties | Low | Part of DevTools, deep integration may depend on Zed's extension API evolution |
| 11 | Performance Overlay | Toggle Flutter performance overlay from the editor | Low | Quick toggle via command palette |
| 12 | Test Runner Integration | Run and debug Flutter/Dart tests | Low | Could leverage Zed's existing test infrastructure |
| 13 | Code Snippets | Common Flutter widget snippets (StatelessWidget, StatefulWidget, etc.) | Low | Nice to have, low effort |
| 14 | Flutter Create | Scaffold new Flutter projects from within Zed | Low | Via command palette |

I'll start working on the high-priority items first and iterate from there. Some features (like the device selector UI) will depend on how far Zed's extension API allows customization — I'll adapt based on what's feasible.

Feedback and suggestions welcome!

**Comment by mambucodev** [comment]
Following up on the roadmap above — I've been working on a Flutter extension and wanted to figure out which features are actually doable with the current extension API (v0.7.0). Turns out most of the important stuff is blocked.

### Blocked entirely

- **Hot Reload / Hot Restart** — there are no file-save hooks or editor event callbacks in the extension API. Extensions can't observe saves, so auto-hot-reload isn't possible. There's also no way to register custom commands or communicate with a running process. See [zed#11107](https://github.com/zed-industries/zed/issues/11107) and [zed#18043](https://github.com/zed-industries/zed/issues/18043).
- **Device Manager** — needs some kind of UI (dropdown, panel, status bar item). The extension API has no UI contribution points at all — no panels, no sidebars, nothing. [extensions#1288](https://github.com/zed-industries/extensions/issues/1288) was opened specifically about this (mentions Flutter device picker as a use case). Also [zed#21208](https://github.com/zed-industries/zed/issues/21208).
- **Widget Inspector / Performance Overlay** — same UI problem, plus you'd need a persistent connection to the Flutter VM service which extensions can't maintain.
- **Code Snippets** — no snippet registration API. The snippet system is user-configured via JSON files, extensions can't contribute to it.

### Partially doable (but limited)

- **Run/Launch** — DAP works, and the Dart extension already hooks up `flutter debug_adapter`. But there's no way to do a plain `flutter run` without the debugger, and extensions can't register tasks ([zed#47068](https://github.com/zed-industries/zed/issues/47068) — LSP tasks are hardcoded to rust-analyzer).
- **Flutter Commands** (`doctor`, `clean`, `pub get`) — you can run these via `Command::output()` but only from slash commands in the AI assistant. No way to put them in the command palette or task list.
- **DevTools** — could technically exec `xdg-open` to open it in a browser, but it depends on WASM sandbox permissions and there's no proper `open_url()` API.
- **Test Runner** — DAP can launch tests, but there's no test results panel or gutter indicators. [zed#5242](https://github.com/zed-industries/zed/issues/5242) tracks a proper test runner.
- **SDK Path / Settings** — can read env vars and `which("flutter")`, but extensions can't define custom settings schemas.
- **Run Configurations** — users can write `.zed/debug.json` manually, but extensions can't generate or manage these.

### Already covered

**DAP Debugging** is handled by the [Dart extension](https://github.com/zed-extensions/dart) — breakpoints, variable inspection, call stacks all work through `flutter debug_adapter`.

### The core gaps

Most of this comes down to a few missing pieces in the extension API:

1. No editor events (save, open, change) — [zed#11107](https://github.com/zed-industries/zed/issues/11107)
2. No custom UI — [extensions#1288](https://github.com/zed-industries/extensions/issues/1288), [zed#21208](https://github.com/zed-industries/zed/issues/21208)
3. No command palette registration — [zed#18043](https://github.com/zed-industries/zed/issues/18043)
4. No task provider for extensions — [zed#47068](https://github.com/zed-industries/zed/issues/47068)
5. No custom LSP message passthrough — the Dart analysis server has custom messages for Flutter features (devices, hot reload, etc.) but extensions can't send/receive them ([zed#20042](https://github.com/zed-industries/zed/issues/20042), [zed#13756](https://github.com/zed-industries/zed/issues/13756))

Right now a Flutter extension can really only do slash commands and piggyback on the Dart extension's DAP support. The IDE tooling people actually want is blocked upstream.

**Comment by vleeuwenmenno** [comment]
Be careful overusing AI slop responses, please read through the relevant issues, most of these findings were already present and people were already aware of them.

**Comment by chandram-adora** [comment]
as long as zed adds supports for ui elements via extension, custom scripts for building and running(tasks through extensions) and integrating flutter daemon via extension, these issues will be blocked.

**Comment by chandram-adora** [comment]
hot reload and restart can be hooked into the existing format on save feature that zed has, that would be a good starting point imo

**Comment by mambucodev** [comment]
> Be careful overusing AI slop responses, please read through the relevant issues, most of these findings were already present and people were already aware of them.

yeah I know, I just used AI for research between this whole repo's issues and also zed's repo, and to reply. I just wanted to recap even what people where already aware about, sorry if it was an inconvenience. 

In fact I think I got to the conclusion that making a Flutter related extension is pretty useless, as debugging and other cool features are still not allowed by zed's extensions api and therefore cannot be integrated.

**Comment by nelsoncampos-cloudwalk** [comment]
Building on @mambucodev's analysis, I dug into the Zed internals to see what the smallest extension API change would be to unblock native hot reload and restart on file save.

Two additions to the WIT interface would cover it:

```wit
// dap.wit, extension calls this to send a custom request to a running session
send-custom-dap-request: func(adapter-name: string, command: string, args: string) -> result<string, string>;
```

```wit
// extension.wit, Zed calls this when a file is saved during a debug session
export on-document-saved: func(adapter-name: string, path: string);
```

Neither is Flutter-specific. Any adapter with custom requests would benefit.

Happy to put together a PR if this direction makes sense. Would love to hear if there's a preferred approach from the Zed side.

**Comment by lordersaur** [comment]
Any updates on this? @nelsoncampos-cloudwalk your PR could change lives! 😂

**Comment by mathis6787** [comment]
I hope it will get implemented soon, zed seems amazing ! I am waiting for this to work on zed Serverpod extension and Jaspr.

**Comment by Aashish-fynd** [comment]
Any updates on this ? I really want to avoid switching to other ides just for getting flutter support.

**Comment by erlangparasu** [comment]
most wanted https://github.com/zed-industries/extensions/issues?q=is%3Aissue%20state%3Aopen%20sort%3Areactions-%2B1-desc

**Comment by gbtb16** [comment]
if zed is able to update what mambuco said, it will be pretty amazing. flutter support is the only reason to use vscode in my workflow yet...

**Comment by van-sprundel** [comment]
AFAIK hot reloading is not part of the DAP right now, so we would need to support custom requests. I'm sure other extensions would benefit from that anyway

https://github.com/zed-industries/zed/blob/94c6647995181ea5e5f4888b4dda2f6d078e78b5/crates/dap/src/client.rs#L98-L100

**Comment by mambucodev** [comment]
My take on this is: there's a lot of core functionalities still missing. The extensions API is still very limited, and some features are even missing from Zed itself.
Last time I used Zed (not long ago) there were no way to run any code block (not just Flutter) from the IDE itself, which would be a really important functionality for this extension to become real.

It's not just about Flutter, even though it's a great example as it requires a broad implementation inside the IDE you're using. Other languages need this too.

I take VSCode and IntelliJ as good references to demonstrate how they have very good implementations of programming languages inside their UIs. Being able to run the code without using a terminal (a simple Run button) is an example of how immersive they can be. Same about Hot reloading, Device management, and so on.

A run widget wouldn't be so bad, even less if it had a wide extension API implementation, so that Flutter or Kotlin apps could eventually show a device list without needing Zed to support it natively.
Or the possibility to have widgets in general, inside of the IDE. That would allow us to have Flutter DevTools, an interactive preview of the app, we could do so many things if we had those functionalities.

I'm not sure if this is on Zed's roadmap, but it would be one of the many steps needed to get to what we need for a Flutter implementation, and many other things.

Please correct me if anything I've said is inaccurate. This comment won't solve or fix anything, just acknowledge that we're far from having a full implementation if the core features we need are not even existent.

**Comment by justkelvin** [comment]
yeah, zed is not ready for flutter support despite them saying Dart is fully supported, I think they don't have or plan on extending what they already have sadly, I use zed as a daily driver for my dev work and for flutter I run everything from cli, already used to it.

**Comment by DylanBolin42** [comment]
Therefore, I look forward to zed opening up more APIs to provide a foundation for writing Flutter-supported plugins.

**Comment by MrSubidubi** [comment]
To provide some context for this (and ping too many people with, apologies for that) - there was an attempt to extend the API recently which I had to close yesterday due to missing context, please see https://github.com/zed-industries/zed/pull/51873#issuecomment-5344990717 for the reasoning behind this, I don't think copying the response here adds any value. 

Also please do note that we already provide some limited support through the Dart extension as mentioned some time within the thread, just for completeness if someone stumbles across this here through searching.

However, I also want to reply to some of the stuff posted here. To quote
> No task provider for extensions

is not entirely true, there is already support for providing tasks through a `tasks.json` and and also ongoing work in the main repo to expose that programatically.

---

But more importantly,
> they don't have or plan on extending what they already have 

is not at all true - we are a code editor after all and our goal is to support code editing in the best and nicest way possible, for example by having programatically defined tasks. Do we hit that mark currenty? Surely not, and surely there are more than rough corners, but that does not mean at all that we are not interested in improving or rewriting the core structure to better support this. Matter of fact, I'd very much like it to.

At the same time, we currently definitely don't expand as much in that area as we probably could, which mostly happens because we spend most time reviewing and trying to get a hold of that. We value all your contributions and try to get back there, it's a tradeoff really of what to do in such case.

What that definitely leads to, though, is that we prefer to look into either open PRs or scope our work to features where it is clear what needs to be done. Supporting Flutter workflows well does sadly not fall in these categories - not because it is not clear what you guys need here, but because we need to both find and research ways that support your use case whilst also working for most to optimally all the languages, DAPs and whatnot out there. We can make the Flutter experience great, sure, but then there will be some other people with another language and we just ruined theirs because of what we did if we follow down that path.

So, in that sense, the issue here is not what to do for flutter, the issue is what can we add for all languages in the API that also works for Flutter. For that, as mentioned in the linked response above, we'd appreciate help, since with all the research we could do we will probably miss something someone in this thread is well aware of.

In that sense - please put the knowledge in **bundled** discussions in the main repository and let us go from there, please. Gathering information there is great, we can get back there and steer, but most importantly, we can land on an implementation that we can be confident about will work for the most cases and not just that one presented here. That is way more actionable for us than issues and also much better a point to steer contributions from. I have reviewed plenty of extensions and thus will have context in areas you might not have, yet every one of you probably has more experience in different areas or knows something that I do not know. Gathering that knowledge in one discussion, agreeing on a path there and then pursuing that path is what I'd consider time best spent on. Posting yet another comment here asking how it can be that this still has not been implemented, not so much. Thank you.

**Comment by c3k4ah** [comment]
Hi everyone, I created this little [Flutter & AVD Manager](https://github.com/c3k4ah/flutter-avd-manager) to make things easier for me when working on Flutter projects with Zed. I've only been using Zed for a short time, and managing AVDs and Flutter features was a real headache. It's not very stable yet, but it does the job for now. It's not directly related to Zed IDE, but at least you can use it without losing hot reload/restart experience, even without Zed extensions.
